### ISSUE TITLE
cmd/go: allow versioned `go build` commands

### ISSUE THREAD

**Issue by hyangah** [golang/go#44469]
Alternative: allow to change the installed main package name in `go install` command

Since go1.16, `go install` accepts arguments with version suffixes. `go install` places the output in `$GOPATH/bin` or `$GOBIN`, but does not allow the output binary's name change.

When a user wants to use multiple versions of tools, the user need to be careful not to overwrite other versions.

Different versions may be coming from different module paths.
e.g. VS Code Go extension uses two different versions of `gocode` (stamblerre/gocode and mdempsky/gocode). In order to install the former, it does switching `GOBIN` to use a temporary directory, installing it, and renaming and copying it to the real `GOBIN`.

There are other proposals for applying the same to `go run` (https://github.com/golang/go/issues/42088) and `go list`(https://github.com/golang/go/issues/44203). Extending `go build` in the same way (or extending `go install`) will help simplifying the tool installation workflow.

cc @jayconrod @bcmills @matloob

**Comment by mvdan** [comment]
On #44203, @jayconrod said:

> Though I think we should talk about allowing `@version` on all commands that take packages (where it makes sense; not `go generate`).

If the plan is indeed to support versioned variants of multiple commands, perhaps we should use a single issue for all of them at once instead of one per command.

**Comment by bcmills** [comment]
> Alternative: allow to change the installed main package name in `go install` command

I think we should _definitely_ do that. `go install` ought to support the same `-o` flag as `go build` and `go test -c`.

Can we make this issue about that, and discuss generalizing `@version` support in a separate issue?

**Comment by hyangah** [comment]
Assuming the scope of #44203 is not limited to `go list` but all applicable commands including `go build`, I am fine with repurposing this issue to discuss the possibility of adding `-o` to `go install`.

What will be the expected behavior of `-o`? For my use case, I hope to see 

`go install -o gopls-master golang.org/x/tools/gopls@master` 

which installs the output in `$GOBIN/gopls-master` (or `$GOBIN\gopls-master.exe`). 
That can be achievable with `go build -o "$(go env GOBIN)/gopls-master" golang.org/x/tools/gopls@master` in _some platforms_ if go build has version support, but still not as convenient as the above command.

**Comment by bcmills** [comment]
I would expect `-o` to mean exactly the same thing for `go install` that it means for `go build`. So you would still need the command to be something like:
```
$ go install -o $(go env GOBIN)/gopls-master golang.org/x/tools/gopls@master
```

Although honestly, in my own use I would probably spell that as:
```
$ go install -o ~bin/gopls-master golang.org/x/tools/gopls@master
```

**Comment by mvdan** [comment]
> go install -o $(go env GOBIN)/gopls-master

The downside of that is that it depends on POSIX Shell syntax, so it wouldn't be portable to e.g. Windows. I see little benefit of `go install` having a `-o` flag if one can use `go build` in exactly the same way, to be honest.

I do see the point in flags being consistent between commands, but the whole point of `go install` is to put the binary in `GOBIN`, hence the flag could be more of a "rename the binary when installing it". That could always be a separate flag, though.

**Comment by bcmills** [comment]
If there were such a flag, I would not want to name it `-o`. `GOFLAGS=-o=a.out` should mean the same thing whether it is applying to `go build` or `go install`.

**Comment by mvdan** [comment]
Maybe. The flip side of the coin is that `go install -o` would make little sense, since `go build -o` would be much clearer :)

**Comment by hyangah** [comment]
`go build -o foo` leaves the output `foo` in the current directory if foo doesn't include the directory path.
Is it too different if `go install -o foo` installs the main package as foo in the default install directory?

**Comment by bcmills** [comment]
> Is it too different if `go install -o foo` installs the main package as foo in the default install directory?

IMO yes, but I'd like to see what others think. 😅

**Comment by jayconrod** [comment]
I think if `go install` gets an `-o` flag, it would be confusing if it didn't do the same thing as `go build`. If it only changes the file name, it should be called something else. I think it would be better to just have `-o` though.

Agree that it's redundant with `go build -o` if `go build` gets an `@version` form though (which I think is a good idea).

**Comment by hyangah** [comment]
Currently we are using the sequence of 
```
cd $(mktemp -d)
go mod init something
go get -d <tool>
go build -o <destination> <tool>
and cleanup... 
```
`go build -o <destination> tool@version` will definitely help us, but still computing `<destination>` for all platforms needs a bit of work. (checking GOBIN, GOPATH, and adding the right file extension, ...)

**Comment by jayconrod** [comment]
If it helps, `<destination>` can be a directory. The executable will be named after the package in that case, but it will have the `.exe` on Windows at least.

When you say "for all platforms", are you cross-compiling? If `GOOS` or `GOARCH` are set to something other than the default, `go install` puts the result in `${GOBIN}/${GOOS}_${GOARCH}/` instead of `${GOBIN}`.

**Comment by hyangah** [comment]
In our specific case, we don't do cross compiling which simplifies things. 
We will need to rename the executable (e.g. gocode -> gocode-gomod, dlv -> dlv-dap, ...). 
Currently we are using various hacks 
https://github.com/golang/vscode-go/blob/master/build/all.bash#L108-L111 and so on
and also in our js/ts code (where I just found another bug in computing the <destination> directory :-) .

To me `-o` semantic isn't too different from `go build`'s `-o`.

* if it's the full path, leave the binary there
* if it's a binary name (without directory), leave the binary with the name in the default directory. For `go build`, the default directory is the current directory where the command runs. For `go install`, the default directory is the default installation directory determined by `GOBIN`, `GOPATH` (and `GOOS` and `GOARCH` if you cross compile).

**Comment by seankhliao** [comment]
Do I have this right? Though I'm not sure when you'd want `go build pkg@version`

| cmd                             | code           | output           |
| ------------------------------- | -------------- | ---------------- |
| `go build pkg`                  | current module | local directory  |
| `go build -o out pkg`           | current module | explicit path    |
| `go build pkg@version`          | remote code    | local directory  |
| `go build -o out pkg@version`   | remote code    | explicit path    |
| `go install pkg`                | current module | global directory |
| `go install -o out pkg`         | current module | explicit path    |
| `go install pkg@version`        | remote code    | global directory |
| `go install -o out pkg@version` | remote code    | explicit path    |

**Comment by meling** [comment]
IMO, the `go install -o out pkg` should install in the default location as suggested by @hyangah above, if out is just a file name. Otherwise, the build and install cases become the same... And there isn’t a convenient way to install a differently named binary in the default location.

**Comment by bcmills** [comment]
@meling
> Otherwise, the build and install cases become the same... And there isn’t a convenient way to install a differently named binary in the default location.

I would rather we have consistent meanings for flags than make every subcommand completely orthogonal.

It's not obvious to me that “installing a differently named binary in the default location” really _needs_ to be convenient.

**Comment by hyangah** [comment]
Thanks @seankhliao. 
`go install -o out pkg@version` and `go build -o out pkg@version` are indeed same except that - in my opinion, where the binary is installed. If `out` is the full path, yes, they are identical.

Is there any command that can provide the default installation directory? If so, I am happy to use it to formulate the right `go build` command.

**Comment by bcmills** [comment]
> Is there any command that can provide the default installation directory?

Within a module, something like `go get -d golang.org/x/tools/gopls && go list -f {{.Target}} golang.org/x/tools/gopls` would do the trick.

If we accept #44203, that would shorten to `go list -f {{.Target}} golang.org/x/tools/gopls@latest`.

**Comment by rlanhellas** [comment]
Hi folks, any progress here?  Is there a roadmap or something else to implement this feature ?

**Comment by mvdan** [comment]
@rlanhellas subscribe to this thread; you'll get notified when there is an update. This is not in the immediate roadmap, given that we're in a code freeze for 1.18 for at least another month.

**Comment by rlanhellas** [comment]
> @rlanhellas subscribe to this thread; you'll get notified when there is an update. This is not in the immediate roadmap, given that we're in a code freeze for 1.18 for at least another month.

thanks @mvdan , I would like to help implement this feature, can I send a `merge request` or someone is working on it ?

**Comment by mvdan** [comment]
We're not even sure if we want this feature or how it would work, see the previous discussion and the `NeedsInvestigation` label. This is also not a particularly good issue to do your first contribution on, as it's somewhat complex. I'd suggest looking at issues with the labels `help wanted` or `NeedsFix`, which are often better candidates.

**Comment by rlanhellas** [comment]
> We're not even sure if we want this feature or how it would work, see the previous discussion and the `NeedsInvestigation` label. This is also not a particularly good issue to do your first contribution on, as it's somewhat complex. I'd suggest looking at issues with the labels `help wanted` or `NeedsFix`, which are often better candidates.

ok @mvdan , thanks for the feedback.

**Comment by ldemailly** [comment]
Repeating here the problem I raised earlier in the hope to get a solution or traction:

In order to take advantage of the new in 1.18 fantastic `BuildInfo` it's ideal (imo) to replace the before classic best practice:
`go build` with `-ldflags -X main.version=1.2.3` by `go install pkg@v1.2.3`

But `go install` doesn't simply let you specify where to install the binary. It puts it in GOPATH/bin when on the native architecture and GOPATH/bin/$arch/  when not; which makes it hacky and hard to just produce the binaries in a multi arch CIs

Example: before:
```bash
CGO_ENABLED=0 GOOS=windows go build -a -ldflags \
'-s -X fortio.org/fortio/version.version=1.28.0 -X "fortio.org/fortio/version.buildInfo=2022-04-26 00:40 57e9d8f01c342a4c1b5d96f883f81128b04d991a"' \
 -o /tmp/fortio.exe fortio.org/fortio
```
after/now
```bash
GOPATH=/build CGO_ENABLED=0 GOOS=darwin /usr/local/go/bin/go install -a -ldflags -s fortio.org/fortio@v1.29.0-pre16
mv -f /build/bin/*_*/fortio* /build/bin
rmdir /build/bin/*_*
```

I end up having to use GOPATH to try to set where the binary ends up as well as using some hacks to deal with native vs non native arch

**Comment by ahyangyi** [comment]
> @meling
> 
> > Otherwise, the build and install cases become the same... And there isn’t a convenient way to install a differently named binary in the default location.
> 
> I would rather we have consistent meanings for flags than make every subcommand completely orthogonal.
> 
> It's not obvious to me that “installing a differently named binary in the default location” really _needs_ to be convenient.

I'm working on a non-golang project that depends on a handful of golang binaries. Which unfortunately all have the name "cmd". Please enlighten me why I don't need a `go install -o <foo>` command.

Sure, I will work around this issue by creating a new directory, run `go build -o` instead and ask the user to add that directory to `PATH`. But, yeah, extra chore.

And if your concern is consistent meanings, then what about giving it a name that's not `-o`? like `-command-name` or something: when we want to change the name when running `go install`, we are not specifying a new *file name* (e.g. `foo.exe` under Windows), we are specifying a new *command name* (e.g., `foo`). That's self-explanatory enough and clear enough, OK?

**Comment by bcap** [comment]
TL;DR: Usecase showing how name clashing during `go install` will cause silent overwrites of previous tools. That coupled with the lack of ability to change the tool name will lead to some awkward hacks

-------

To leave a practical example of how cumbersome the user experience can be with `go install` today:

Usecase: Run the tool defined at github.com/Jille/raft-grpc-example/cmd/hammer plus some other raft-related tooling

I can go install `github.com/Jille/raft-grpc-example/cmd/hammer`, but then what happens is that will overwrite another `hammer` tool that I already have on my `~/go/bin`. It is name clashing and the overwrite will happen silenty, which is bad. 

Anyway, ideally I would want to build it as "raft-hammer"

So this was my user experience:

Lets try installing it with a different name. How do I do so?

```
% go help install

usage: go install [build flags] [packages]

Install compiles and installs the packages named by the import paths.

...

Ok, so it takes build flags. Lets try:

```
% go install -o raft-hammer github.com/Jille/raft-grpc-example/cmd/hammer@latest
flag provided but not defined: -o
usage: go install [build flags] [packages]
Run 'go help install' for details.
```

Well, thats weird, takes build flags but not _this_ one build flag. Lets google it. I ended up finding this GH Issue

Fair enough, seems like this discussion went a lot around `go install` becoming the same as `go build` and how bad would it be to have more flags on go install, which IMO doesnt hold as a strong argument. It seems like we cant add `-o` to `go install` to avoid commands having specific flags, but `-o` is already special anyway because it is not passed to `go build` by `go install`. So `-o` is basically a flag sitting on some kind of a _limbo_.

Anyway, what if we try to simple `go build` then?

```
go build -o raft-hammer github.com/Jille/raft-grpc-example/cmd/hammer@latest
package github.com/Jille/raft-grpc-example/cmd/hammer@latest: can only use path@version syntax with 'go get' and 'go install' in module-aware mode
```

Ok, so `go build` is not version aware. Not sure why but that is confusing. 

Well I guess we need to go for `go install` + minor hacking:
```
% mv ~/go/bin/{hammer,hammer-swap}
% go install github.com/Jille/raft-grpc-example/cmd/hammer@latest
% mv ~/go/bin/{hammer,raft-hammer}
% mv ~/go/bin/{hammer-swap,hammer}
```

**Comment by oliverpool** [comment]
@bcap I had the same issue, but used the [workaround of @hyangah](https://github.com/golang/go/issues/44469#issuecomment-785246687): add a `go.mod` file.

You can then use `go get` and `go build`:
```
go get github.com/Jille/raft-grpc-example/cmd/hammer@latest
go build -o ~/go/bin/raft-hammer github.com/Jille/raft-grpc-example/cmd/hammer
```
It even works for cross-compilation ([go install](https://github.com/golang/go/issues/57485) currently refuses a custom GOBIN for cross-compilation)

**Comment by lyda** [comment]
Sigh.  It would be really nice if something like this existed.  Reading through the objections, how about using `-O` (an upper-case "O")?

```bash
go install -O foozle some.git.host/flibble,burble/foozle.git@v1.3-kitten
```

That way it doesn't clash with the `-o` flag elsewhere.  Don't allow a slash in `foozle` and put it the same place that `foozle.git` currently goes.  This doesn't solve all problems, but it solves a lot of them.

**Comment by ldemailly** [comment]
being able to specify a full destination path including directory (ie with slashes) is what I need for ci builds or predictable installs but maybe #52898 can be reopen if the direction for this one is to only address the binary basename

**Comment by matloob** [comment]
I think that since we support `go run pkg@version` it makes sense to also support `go build pkg@version` like others have suggested. Though I wonder if we might want to follow @mvdan's suggestion in https://github.com/golang/go/issues/44469#issuecomment-782890190 to make a single issue to add pkg@version to all commands where it makes sense?

**Comment by matloob** [comment]
@hyangah Should we turn this into a proposal issue?

**Comment by hyangah** [comment]
Do we need more than marking this as a proposal?

**Comment by matloob** [comment]
I think we could update the first comment with the current version of the proposal?

I'll mark this as a proposal.

**Comment by eyjian** [comment]
I need "go install" to specify an abbreviation different from the package name, for example: go-effex-cli is gec in my project, supporting "-o" is a good idea.

**Comment by changkun** [comment]
What is blocking here?

**Comment by prattmic** [comment]
I'll add to @ldemailly's cross-compilation use case in https://github.com/golang/go/issues/44469#issuecomment-1439699748.

For issues about telemetry-reported crashes like #75673, we'd like to provide a simple command to reproduce the crashing binary. `go install`'s default output location varying depending on whether this is a cross-compilation or not makes it quite awkward to provide a simple universal command.

`go build path@version` or `go install -o` would both solve this. (In the latter case, provided that `go install -o` outputs to the current directory, not relative to the GOBIN location)

**Comment by matloob** [comment]
We should go ahead with `go build pkg@version`. It's something that we arguably should support even if we end up later adding a `go install -o` option (though I personally don't think we should add `go install -o`).

I think at this point we just need to get the proposal appreved.

**Comment by firelizzard18** [comment]
With support for `go build pkg@version`, is there any difference between `go build -o $PATH pkg@version` and `go install -o $PATH pkg@version`?

**Comment by ldemailly** [comment]
having -o on install being a bit smart, if -o foo -> ./foo if -o ./dir -> ./dir/pkgname etc (and safer obviously with -o /full/path) should disambiguate the issues automatically

this being said now that go build also embbeds buildinfo it has become less necessary - yet being able to say

go install -o /usr/local/bin example.com/superprogram@latest

would be really nice to not have to follow up with "have you put GOBIN in your path" troubleshooting
